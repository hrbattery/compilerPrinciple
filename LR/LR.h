/*************************************************************************
	> File Name: LR.h
	  > Author: Netcan
	  > Blog: http://www.netcan666.com
	  > Mail: 1469709759@qq.com
	  > Created Time: 2016-10-19 三 08:36:05 CST

                                                    __----~~~~~~~~~~~------___
                                   .  .   ~~//====......          __--~ ~~
                   -.            \_|//     |||\\  ~~~~~~::::... /~
                ___-==_       _-~o~  \/    |||  \\            _/~~-
        __---~~~.==~||\=_    -_--~/_-~|-   |\\   \\        _/~
    _-~~     .=~    |  \\-_    '-~7  /-   /  ||    \      /
  .~       .~       |   \\ -_    /  /-   /   ||      \   /
 /  ____  /         |     \\ ~-_/  /|- _/   .||       \ /
 |~~    ~~|--~~~~--_ \     ~==-/   | \~--===~~        .\
          '         ~-|      /|    |-~\~~       __--~~
                      |-~~-_/ |    |   ~\_   _-~            /\
                           /  \     \__   \/~                \__
                       _--~ _/ | .-~~____--~-/                  ~~==.
                      ((->/~   '.|||' -_|    ~~-/ ,              . _||
                                 -_     ~\      ~~---l__i__i__i--~~_/
                                 _-~-__   ~)  \--______________--~~
                               //.-~~~-~_--~- |-------~~~~~~~~
                                      //.-~~~--\
						神兽保佑，代码无BUG!
 ************************************************************************/

#ifndef _LR_H
#define _LR_H

#include <iostream>
#include <ctype.h>
#include <vector>
#include <string>
#include <algorithm>
#include <map>
#include <set>
using namespace std;
#define EXTENSION_NOTERMINAL '^'

class Prod { // 这里是存放形如X->abc的形式，不存在多个候选式
	friend class Item;
	friend class LR;
	private:
		char noTerminal; // 产生式左部非终结符名字
		string right; // 产生式右部
		set<char> additionalVt; // 附加终结符
		friend bool operator == (const Prod &a, const Prod &b) {
			return a.noTerminal == b.noTerminal && a.right == b.right;
		}
		friend bool operator == (const Prod &a, char c) {
			return a.noTerminal == c;
		}

	public:
		static string cut(const string &in, int i, int j) {
			return string(in.begin() + i, in.begin() + j);
		}
		static string replaceAll(const string &in, const string from, const string to);
		string displayStr() const;
		Prod(const string &in);
		Prod(const char &noTerminal, const string& right, const set<char>& additionalVt):
			noTerminal(noTerminal), right(right), additionalVt(additionalVt) {}
};

class Item { // 项目集
	friend class LR;
	private:
		vector<Prod> prods; // 项目集
		static set<char> Vn; // 非终结符
		static set<char> Vt; // 终结符
		static set<char> Symbol; // 所有符号
		friend bool operator == (const Item &a, const Item &b) {
			if(a.prods.size() != b.prods.size()) return false;
			else {
				for(const auto& p: a.prods) {// 选择a的每个产生式
					auto it = find(b.prods.begin(), b.prods.end(), p);
					if(it == b.prods.end())  // 找不到
						return false;
					else  {// 找到的话判断附加终结符是否都相等
						if(p.additionalVt != it->additionalVt)
							return false;
					}
				}
				return true;
			}
		}
	public:
		void add(const string &prod);
		void display() const;
};

class LR {
	private:
		Item G; // 文法G
		enum actionStat{
			ACCEPT=0,
			SHIFT,
			REDUCE,
		};
		static const char *actionStatStr[];

		vector<Item> C; // 项目集规范族
		map<pair<int, char>, int> GOTO; // goto数组，项目集<int, int>=char
		map<pair<int, char>, pair<actionStat, int> > ACTION; // Action数组，Action[(i, a)]=(s|r)j
		map<char, set<char> > FIRST; // first集
		map<char, set<char> > FOLLOW; // follow集
		set<char> first(const string &s); // 求first集
		void follow();
		vector<char> inStr; // 输入串/栈
		vector<int> status; // 状态栈
		vector<char> parse; // 分析栈
		Item closure(Item I); // 求该项目的闭包
		Item Goto(const Item& I, char X); // 求I经过X到达的项目集
		void items(); // 求项目集状态机DFA！!
		void showStrStack(); // 显示输入栈
		void showStatusStack();
		void showParseStack();
	public:
		void add(const string &s); // 添加产生式
		void build(); // 构造Action、GOTO表
		void showTable(); // 打印LR分析表！
		void debug();
		void loadStr(const string &in); // 读取输入串
		void parser(); // LR(1)分析
		void showGrammar(); // 显示输入的文法产生式
		void drawGraph(); // 画出DFA！
		void generateDot();
		void run();
};

#endif
