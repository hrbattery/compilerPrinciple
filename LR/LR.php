<?php
header('Content-Type: application/json');

$data = $_POST['data'] . "\n#\n#";
$descriptorspec = array(
   0 => array("pipe", "r"),  // stdin is a pipe that the child will read from
   1 => array("pipe", "w"),  // stdout is a pipe that the child will write to
);

if (strtoupper(substr(PHP_OS, 0, 3)) === 'WIN')
	$process = proc_open('LR', $descriptorspec, $pipes);
else
	$process = proc_open('./LR', $descriptorspec, $pipes);

if(is_resource($process)) {

	fwrite($pipes[0], $data);
	fclose($pipes[0]);

	echo stream_get_contents($pipes[1]);
	fclose($pipes[1]);
	proc_close($process);
}


