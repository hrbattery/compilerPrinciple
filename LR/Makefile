CXX=clang++
CFLAGS=-O3 --std=c++11

all: LR LR_DFA

LR: LR.cpp
	$(CXX) $(CFLAGS) $< -o $@

LR_DFA: LR_DFA.cpp LR.cpp
	$(CXX) $(CFLAGS) -D_GENERATE_DOT_  $^ -o $@

clean:
	@rm LR LR_DFA

test: LR LR_DFA
	for i in $$(ls Test/*.in); \
	do\
		./LR < $$i > "$$(echo $$i | sed -r 's/(.*)\.in/\1\.out/')";\
		./LR_DFA < $$i | dot -Tpdf -o "$$(echo $$i | sed -r 's/(.*)\.in/\1\.pdf/')";\
	done
