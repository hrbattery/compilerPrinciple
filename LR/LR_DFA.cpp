/*************************************************************************
	> File Name: LR_DFA.cpp
	  > Author: Netcan
	  > Blog: http://www.netcan666.com
	  > Mail: 1469709759@qq.com
	  > Created Time: 2016-10-21 五 23:08:06 CST

                                                    __----~~~~~~~~~~~------___
                                   .  .   ~~//====......          __--~ ~~
                   -.            \_|//     |||\\  ~~~~~~::::... /~
                ___-==_       _-~o~  \/    |||  \\            _/~~-
        __---~~~.==~||\=_    -_--~/_-~|-   |\\   \\        _/~
    _-~~     .=~    |  \\-_    '-~7  /-   /  ||    \      /
  .~       .~       |   \\ -_    /  /-   /   ||      \   /
 /  ____  /         |     \\ ~-_/  /|- _/   .||       \ /
 |~~    ~~|--~~~~--_ \     ~==-/   | \~--===~~        .\
          '         ~-|      /|    |-~\~~       __--~~
                      |-~~-_/ |    |   ~\_   _-~            /\
                           /  \     \__   \/~                \__
                       _--~ _/ | .-~~____--~-/                  ~~==.
                      ((->/~   '.|||' -_|    ~~-/ ,              . _||
                                 -_     ~\      ~~---l__i__i__i--~~_/
                                 _-~-__   ~)  \--______________--~~
                               //.-~~~-~_--~- |-------~~~~~~~~
                                      //.-~~~--\
						神兽保佑，代码无BUG!
 ************************************************************************/

#include "LR.h"

int main() {
	LR lr;
	string in;
	while(cin >> in && in != "#")
		lr.add(in);
	in = "";
	cin >> in;
	if(in == "#" || in.size() == 0) return -1;
	lr.loadStr(in);
	lr.build();
	lr.generateDot();
}
