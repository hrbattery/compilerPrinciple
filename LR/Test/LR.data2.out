{"Grammar": [
 "E->TG", "G->+TG", "G->-TG", "G->@", "T->FS", "S->*FS", "S->/FS", "S->@", "F->(E)", "F->i"
]
,"parseTable": {
"Vt": ["(", ")", "*", "+", "-", "/", "i", "#"],
"Vn": [ "E", "F", "G", "S", "T"],
"Body": [
 [ "s1", "", "", "", "", "", "s5", "", 2, 3, "", "", 4]
, [ "s6", "", "", "", "", "", "s10", "", 7, 8, "", "", 9]
, [ "", "", "", "", "", "", "", "acc", "", "", "", "", ""]
, [ "", "", "s11", "r7", "r7", "s12", "", "r7", "", "", "", 13, ""]
, [ "", "", "", "s14", "s15", "", "", "r3", "", "", 16, "", ""]
, [ "", "", "r9", "r9", "r9", "r9", "", "r9", "", "", "", "", ""]
, [ "s6", "", "", "", "", "", "s10", "", 17, 8, "", "", 9]
, [ "", "s18", "", "", "", "", "", "", "", "", "", "", ""]
, [ "", "r7", "s19", "r7", "r7", "s20", "", "", "", "", "", 21, ""]
, [ "", "r3", "", "s22", "s23", "", "", "", "", "", 24, "", ""]
, [ "", "r9", "r9", "r9", "r9", "r9", "", "", "", "", "", "", ""]
, [ "s1", "", "", "", "", "", "s5", "", "", 25, "", "", ""]
, [ "s1", "", "", "", "", "", "s5", "", "", 26, "", "", ""]
, [ "", "", "", "r4", "r4", "", "", "r4", "", "", "", "", ""]
, [ "s1", "", "", "", "", "", "s5", "", "", 3, "", "", 27]
, [ "s1", "", "", "", "", "", "s5", "", "", 3, "", "", 28]
, [ "", "", "", "", "", "", "", "r0", "", "", "", "", ""]
, [ "", "s29", "", "", "", "", "", "", "", "", "", "", ""]
, [ "", "", "r8", "r8", "r8", "r8", "", "r8", "", "", "", "", ""]
, [ "s6", "", "", "", "", "", "s10", "", "", 30, "", "", ""]
, [ "s6", "", "", "", "", "", "s10", "", "", 31, "", "", ""]
, [ "", "r4", "", "r4", "r4", "", "", "", "", "", "", "", ""]
, [ "s6", "", "", "", "", "", "s10", "", "", 8, "", "", 32]
, [ "s6", "", "", "", "", "", "s10", "", "", 8, "", "", 33]
, [ "", "r0", "", "", "", "", "", "", "", "", "", "", ""]
, [ "", "", "s11", "r7", "r7", "s12", "", "r7", "", "", "", 34, ""]
, [ "", "", "s11", "r7", "r7", "s12", "", "r7", "", "", "", 35, ""]
, [ "", "", "", "s14", "s15", "", "", "r3", "", "", 36, "", ""]
, [ "", "", "", "s14", "s15", "", "", "r3", "", "", 37, "", ""]
, [ "", "r8", "r8", "r8", "r8", "r8", "", "", "", "", "", "", ""]
, [ "", "r7", "s19", "r7", "r7", "s20", "", "", "", "", "", 38, ""]
, [ "", "r7", "s19", "r7", "r7", "s20", "", "", "", "", "", 39, ""]
, [ "", "r3", "", "s22", "s23", "", "", "", "", "", 40, "", ""]
, [ "", "r3", "", "s22", "s23", "", "", "", "", "", 41, "", ""]
, [ "", "", "", "r5", "r5", "", "", "r5", "", "", "", "", ""]
, [ "", "", "", "r6", "r6", "", "", "r6", "", "", "", "", ""]
, [ "", "", "", "", "", "", "", "r1", "", "", "", "", ""]
, [ "", "", "", "", "", "", "", "r2", "", "", "", "", ""]
, [ "", "r5", "", "r5", "r5", "", "", "", "", "", "", "", ""]
, [ "", "r6", "", "r6", "r6", "", "", "", "", "", "", "", ""]
, [ "", "r1", "", "", "", "", "", "", "", "", "", "", ""]
, [ "", "r2", "", "", "", "", "", "", "", "", "", "", ""]
]}
,"Graph": {"All": "digraph all{node [shape=box style=filled];Grammar[style=rounded label=\"Grammar\nE->TG\nG->+TG\nG->-TG\nG->@\nT->FS\nS->*FS\nS->/FS\nS->@\nF->(E)\nF->i\n\" ];\nI0[label=\"I0\nE'->.E,#\nE->.TG,#\nT->.FS,+|-|#\nF->.(E),*|+|-|/|#\nF->.i,*|+|-|/|#\n\" ];\nI1[label=\"I1\nF->(.E),*|+|-|/|#\nE->.TG,)\nT->.FS,)|+|-\nF->.(E),)|*|+|-|/\nF->.i,)|*|+|-|/\n\" ];\nI2[label=\"I2\nE'->E.,#\n\" ];\nI3[label=\"I3\nT->F.S,+|-|#\nS->.*FS,+|-|#\nS->./FS,+|-|#\nS->@.,+|-|#\n\" ];\nI4[label=\"I4\nE->T.G,#\nG->.+TG,#\nG->.-TG,#\nG->@.,#\n\" ];\nI5[label=\"I5\nF->i.,*|+|-|/|#\n\" ];\nI6[label=\"I6\nF->(.E),)|*|+|-|/\nE->.TG,)\nT->.FS,)|+|-\nF->.(E),)|*|+|-|/\nF->.i,)|*|+|-|/\n\" ];\nI7[label=\"I7\nF->(E.),*|+|-|/|#\n\" ];\nI8[label=\"I8\nT->F.S,)|+|-\nS->.*FS,)|+|-\nS->./FS,)|+|-\nS->@.,)|+|-\n\" ];\nI9[label=\"I9\nE->T.G,)\nG->.+TG,)\nG->.-TG,)\nG->@.,)\n\" ];\nI10[label=\"I10\nF->i.,)|*|+|-|/\n\" ];\nI11[label=\"I11\nS->*.FS,+|-|#\nF->.(E),*|+|-|/|#\nF->.i,*|+|-|/|#\n\" ];\nI12[label=\"I12\nS->/.FS,+|-|#\nF->.(E),*|+|-|/|#\nF->.i,*|+|-|/|#\n\" ];\nI13[label=\"I13\nT->FS.,+|-|#\n\" ];\nI14[label=\"I14\nG->+.TG,#\nT->.FS,+|-|#\nF->.(E),*|+|-|/|#\nF->.i,*|+|-|/|#\n\" ];\nI15[label=\"I15\nG->-.TG,#\nT->.FS,+|-|#\nF->.(E),*|+|-|/|#\nF->.i,*|+|-|/|#\n\" ];\nI16[label=\"I16\nE->TG.,#\n\" ];\nI17[label=\"I17\nF->(E.),)|*|+|-|/\n\" ];\nI18[label=\"I18\nF->(E).,*|+|-|/|#\n\" ];\nI19[label=\"I19\nS->*.FS,)|+|-\nF->.(E),)|*|+|-|/\nF->.i,)|*|+|-|/\n\" ];\nI20[label=\"I20\nS->/.FS,)|+|-\nF->.(E),)|*|+|-|/\nF->.i,)|*|+|-|/\n\" ];\nI21[label=\"I21\nT->FS.,)|+|-\n\" ];\nI22[label=\"I22\nG->+.TG,)\nT->.FS,)|+|-\nF->.(E),)|*|+|-|/\nF->.i,)|*|+|-|/\n\" ];\nI23[label=\"I23\nG->-.TG,)\nT->.FS,)|+|-\nF->.(E),)|*|+|-|/\nF->.i,)|*|+|-|/\n\" ];\nI24[label=\"I24\nE->TG.,)\n\" ];\nI25[label=\"I25\nS->*F.S,+|-|#\nS->.*FS,+|-|#\nS->./FS,+|-|#\nS->@.,+|-|#\n\" ];\nI26[label=\"I26\nS->/F.S,+|-|#\nS->.*FS,+|-|#\nS->./FS,+|-|#\nS->@.,+|-|#\n\" ];\nI27[label=\"I27\nG->+T.G,#\nG->.+TG,#\nG->.-TG,#\nG->@.,#\n\" ];\nI28[label=\"I28\nG->-T.G,#\nG->.+TG,#\nG->.-TG,#\nG->@.,#\n\" ];\nI29[label=\"I29\nF->(E).,)|*|+|-|/\n\" ];\nI30[label=\"I30\nS->*F.S,)|+|-\nS->.*FS,)|+|-\nS->./FS,)|+|-\nS->@.,)|+|-\n\" ];\nI31[label=\"I31\nS->/F.S,)|+|-\nS->.*FS,)|+|-\nS->./FS,)|+|-\nS->@.,)|+|-\n\" ];\nI32[label=\"I32\nG->+T.G,)\nG->.+TG,)\nG->.-TG,)\nG->@.,)\n\" ];\nI33[label=\"I33\nG->-T.G,)\nG->.+TG,)\nG->.-TG,)\nG->@.,)\n\" ];\nI34[label=\"I34\nS->*FS.,+|-|#\n\" ];\nI35[label=\"I35\nS->/FS.,+|-|#\n\" ];\nI36[label=\"I36\nG->+TG.,#\n\" ];\nI37[label=\"I37\nG->-TG.,#\n\" ];\nI38[label=\"I38\nS->*FS.,)|+|-\n\" ];\nI39[label=\"I39\nS->/FS.,)|+|-\n\" ];\nI40[label=\"I40\nG->+TG.,)\n\" ];\nI41[label=\"I41\nG->-TG.,)\n\" ];\nI0 -> I1[label=\"(\"];I0 -> I2[label=\"E\"];I0 -> I3[label=\"F\"];I0 -> I4[label=\"T\"];I0 -> I5[label=\"i\"];I1 -> I6[label=\"(\"];I1 -> I7[label=\"E\"];I1 -> I8[label=\"F\"];I1 -> I9[label=\"T\"];I1 -> I10[label=\"i\"];I3 -> I11[label=\"*\"];I3 -> I12[label=\"/\"];I3 -> I13[label=\"S\"];I4 -> I14[label=\"+\"];I4 -> I15[label=\"-\"];I4 -> I16[label=\"G\"];I6 -> I6[label=\"(\"];I6 -> I17[label=\"E\"];I6 -> I8[label=\"F\"];I6 -> I9[label=\"T\"];I6 -> I10[label=\"i\"];I7 -> I18[label=\")\"];I8 -> I19[label=\"*\"];I8 -> I20[label=\"/\"];I8 -> I21[label=\"S\"];I9 -> I22[label=\"+\"];I9 -> I23[label=\"-\"];I9 -> I24[label=\"G\"];I11 -> I1[label=\"(\"];I11 -> I25[label=\"F\"];I11 -> I5[label=\"i\"];I12 -> I1[label=\"(\"];I12 -> I26[label=\"F\"];I12 -> I5[label=\"i\"];I14 -> I1[label=\"(\"];I14 -> I3[label=\"F\"];I14 -> I27[label=\"T\"];I14 -> I5[label=\"i\"];I15 -> I1[label=\"(\"];I15 -> I3[label=\"F\"];I15 -> I28[label=\"T\"];I15 -> I5[label=\"i\"];I17 -> I29[label=\")\"];I19 -> I6[label=\"(\"];I19 -> I30[label=\"F\"];I19 -> I10[label=\"i\"];I20 -> I6[label=\"(\"];I20 -> I31[label=\"F\"];I20 -> I10[label=\"i\"];I22 -> I6[label=\"(\"];I22 -> I8[label=\"F\"];I22 -> I32[label=\"T\"];I22 -> I10[label=\"i\"];I23 -> I6[label=\"(\"];I23 -> I8[label=\"F\"];I23 -> I33[label=\"T\"];I23 -> I10[label=\"i\"];I25 -> I11[label=\"*\"];I25 -> I12[label=\"/\"];I25 -> I34[label=\"S\"];I26 -> I11[label=\"*\"];I26 -> I12[label=\"/\"];I26 -> I35[label=\"S\"];I27 -> I14[label=\"+\"];I27 -> I15[label=\"-\"];I27 -> I36[label=\"G\"];I28 -> I14[label=\"+\"];I28 -> I15[label=\"-\"];I28 -> I37[label=\"G\"];I30 -> I19[label=\"*\"];I30 -> I20[label=\"/\"];I30 -> I38[label=\"S\"];I31 -> I19[label=\"*\"];I31 -> I20[label=\"/\"];I31 -> I39[label=\"S\"];I32 -> I22[label=\"+\"];I32 -> I23[label=\"-\"];I32 -> I40[label=\"G\"];I33 -> I22[label=\"+\"];I33 -> I23[label=\"-\"];I33 -> I41[label=\"G\"];Grammar -> I0[style=invis];}"
, "Simple": "digraph simple {node [shape = circle style=filled];Grammar[shape=box style=rounded label=\"Grammar\nE->TG\nG->+TG\nG->-TG\nG->@\nT->FS\nS->*FS\nS->/FS\nS->@\nF->(E)\nF->i\n\" ];\nI0[tooltip=\"I0\nE'->.E,#\nE->.TG,#\nT->.FS,+|-|#\nF->.(E),*|+|-|/|#\nF->.i,*|+|-|/|#\n\" ];\nI1[tooltip=\"I1\nF->(.E),*|+|-|/|#\nE->.TG,)\nT->.FS,)|+|-\nF->.(E),)|*|+|-|/\nF->.i,)|*|+|-|/\n\" ];\nI2[tooltip=\"I2\nE'->E.,#\n\" ];\nI3[tooltip=\"I3\nT->F.S,+|-|#\nS->.*FS,+|-|#\nS->./FS,+|-|#\nS->@.,+|-|#\n\" ];\nI4[tooltip=\"I4\nE->T.G,#\nG->.+TG,#\nG->.-TG,#\nG->@.,#\n\" ];\nI5[tooltip=\"I5\nF->i.,*|+|-|/|#\n\" ];\nI6[tooltip=\"I6\nF->(.E),)|*|+|-|/\nE->.TG,)\nT->.FS,)|+|-\nF->.(E),)|*|+|-|/\nF->.i,)|*|+|-|/\n\" ];\nI7[tooltip=\"I7\nF->(E.),*|+|-|/|#\n\" ];\nI8[tooltip=\"I8\nT->F.S,)|+|-\nS->.*FS,)|+|-\nS->./FS,)|+|-\nS->@.,)|+|-\n\" ];\nI9[tooltip=\"I9\nE->T.G,)\nG->.+TG,)\nG->.-TG,)\nG->@.,)\n\" ];\nI10[tooltip=\"I10\nF->i.,)|*|+|-|/\n\" ];\nI11[tooltip=\"I11\nS->*.FS,+|-|#\nF->.(E),*|+|-|/|#\nF->.i,*|+|-|/|#\n\" ];\nI12[tooltip=\"I12\nS->/.FS,+|-|#\nF->.(E),*|+|-|/|#\nF->.i,*|+|-|/|#\n\" ];\nI13[tooltip=\"I13\nT->FS.,+|-|#\n\" ];\nI14[tooltip=\"I14\nG->+.TG,#\nT->.FS,+|-|#\nF->.(E),*|+|-|/|#\nF->.i,*|+|-|/|#\n\" ];\nI15[tooltip=\"I15\nG->-.TG,#\nT->.FS,+|-|#\nF->.(E),*|+|-|/|#\nF->.i,*|+|-|/|#\n\" ];\nI16[tooltip=\"I16\nE->TG.,#\n\" ];\nI17[tooltip=\"I17\nF->(E.),)|*|+|-|/\n\" ];\nI18[tooltip=\"I18\nF->(E).,*|+|-|/|#\n\" ];\nI19[tooltip=\"I19\nS->*.FS,)|+|-\nF->.(E),)|*|+|-|/\nF->.i,)|*|+|-|/\n\" ];\nI20[tooltip=\"I20\nS->/.FS,)|+|-\nF->.(E),)|*|+|-|/\nF->.i,)|*|+|-|/\n\" ];\nI21[tooltip=\"I21\nT->FS.,)|+|-\n\" ];\nI22[tooltip=\"I22\nG->+.TG,)\nT->.FS,)|+|-\nF->.(E),)|*|+|-|/\nF->.i,)|*|+|-|/\n\" ];\nI23[tooltip=\"I23\nG->-.TG,)\nT->.FS,)|+|-\nF->.(E),)|*|+|-|/\nF->.i,)|*|+|-|/\n\" ];\nI24[tooltip=\"I24\nE->TG.,)\n\" ];\nI25[tooltip=\"I25\nS->*F.S,+|-|#\nS->.*FS,+|-|#\nS->./FS,+|-|#\nS->@.,+|-|#\n\" ];\nI26[tooltip=\"I26\nS->/F.S,+|-|#\nS->.*FS,+|-|#\nS->./FS,+|-|#\nS->@.,+|-|#\n\" ];\nI27[tooltip=\"I27\nG->+T.G,#\nG->.+TG,#\nG->.-TG,#\nG->@.,#\n\" ];\nI28[tooltip=\"I28\nG->-T.G,#\nG->.+TG,#\nG->.-TG,#\nG->@.,#\n\" ];\nI29[tooltip=\"I29\nF->(E).,)|*|+|-|/\n\" ];\nI30[tooltip=\"I30\nS->*F.S,)|+|-\nS->.*FS,)|+|-\nS->./FS,)|+|-\nS->@.,)|+|-\n\" ];\nI31[tooltip=\"I31\nS->/F.S,)|+|-\nS->.*FS,)|+|-\nS->./FS,)|+|-\nS->@.,)|+|-\n\" ];\nI32[tooltip=\"I32\nG->+T.G,)\nG->.+TG,)\nG->.-TG,)\nG->@.,)\n\" ];\nI33[tooltip=\"I33\nG->-T.G,)\nG->.+TG,)\nG->.-TG,)\nG->@.,)\n\" ];\nI34[tooltip=\"I34\nS->*FS.,+|-|#\n\" ];\nI35[tooltip=\"I35\nS->/FS.,+|-|#\n\" ];\nI36[tooltip=\"I36\nG->+TG.,#\n\" ];\nI37[tooltip=\"I37\nG->-TG.,#\n\" ];\nI38[tooltip=\"I38\nS->*FS.,)|+|-\n\" ];\nI39[tooltip=\"I39\nS->/FS.,)|+|-\n\" ];\nI40[tooltip=\"I40\nG->+TG.,)\n\" ];\nI41[tooltip=\"I41\nG->-TG.,)\n\" ];\nI0 -> I1[label=\"(\"];I0 -> I2[label=\"E\"];I0 -> I3[label=\"F\"];I0 -> I4[label=\"T\"];I0 -> I5[label=\"i\"];I1 -> I6[label=\"(\"];I1 -> I7[label=\"E\"];I1 -> I8[label=\"F\"];I1 -> I9[label=\"T\"];I1 -> I10[label=\"i\"];I3 -> I11[label=\"*\"];I3 -> I12[label=\"/\"];I3 -> I13[label=\"S\"];I4 -> I14[label=\"+\"];I4 -> I15[label=\"-\"];I4 -> I16[label=\"G\"];I6 -> I6[label=\"(\"];I6 -> I17[label=\"E\"];I6 -> I8[label=\"F\"];I6 -> I9[label=\"T\"];I6 -> I10[label=\"i\"];I7 -> I18[label=\")\"];I8 -> I19[label=\"*\"];I8 -> I20[label=\"/\"];I8 -> I21[label=\"S\"];I9 -> I22[label=\"+\"];I9 -> I23[label=\"-\"];I9 -> I24[label=\"G\"];I11 -> I1[label=\"(\"];I11 -> I25[label=\"F\"];I11 -> I5[label=\"i\"];I12 -> I1[label=\"(\"];I12 -> I26[label=\"F\"];I12 -> I5[label=\"i\"];I14 -> I1[label=\"(\"];I14 -> I3[label=\"F\"];I14 -> I27[label=\"T\"];I14 -> I5[label=\"i\"];I15 -> I1[label=\"(\"];I15 -> I3[label=\"F\"];I15 -> I28[label=\"T\"];I15 -> I5[label=\"i\"];I17 -> I29[label=\")\"];I19 -> I6[label=\"(\"];I19 -> I30[label=\"F\"];I19 -> I10[label=\"i\"];I20 -> I6[label=\"(\"];I20 -> I31[label=\"F\"];I20 -> I10[label=\"i\"];I22 -> I6[label=\"(\"];I22 -> I8[label=\"F\"];I22 -> I32[label=\"T\"];I22 -> I10[label=\"i\"];I23 -> I6[label=\"(\"];I23 -> I8[label=\"F\"];I23 -> I33[label=\"T\"];I23 -> I10[label=\"i\"];I25 -> I11[label=\"*\"];I25 -> I12[label=\"/\"];I25 -> I34[label=\"S\"];I26 -> I11[label=\"*\"];I26 -> I12[label=\"/\"];I26 -> I35[label=\"S\"];I27 -> I14[label=\"+\"];I27 -> I15[label=\"-\"];I27 -> I36[label=\"G\"];I28 -> I14[label=\"+\"];I28 -> I15[label=\"-\"];I28 -> I37[label=\"G\"];I30 -> I19[label=\"*\"];I30 -> I20[label=\"/\"];I30 -> I38[label=\"S\"];I31 -> I19[label=\"*\"];I31 -> I20[label=\"/\"];I31 -> I39[label=\"S\"];I32 -> I22[label=\"+\"];I32 -> I23[label=\"-\"];I32 -> I40[label=\"G\"];I33 -> I22[label=\"+\"];I33 -> I23[label=\"-\"];I33 -> I41[label=\"G\"];Grammar -> I0[style=invis];}"},
"parser": [
{"statusStack": " 0 ", "parseStack": "", "inStrStack": "i+i*i#", "action": "SHIFT"}
, {"statusStack": " 0  5 ", "parseStack": "i", "inStrStack": "+i*i#", "action": "REDUCE F->i"}
, {"statusStack": " 0  3 ", "parseStack": "F", "inStrStack": "+i*i#", "action": "REDUCE S->@"}
, {"statusStack": " 0  3  <span class='underline'>13</span> ", "parseStack": "FS", "inStrStack": "+i*i#", "action": "REDUCE T->FS"}
, {"statusStack": " 0  4 ", "parseStack": "T", "inStrStack": "+i*i#", "action": "SHIFT"}
, {"statusStack": " 0  4  <span class='underline'>14</span> ", "parseStack": "T+", "inStrStack": "i*i#", "action": "SHIFT"}
, {"statusStack": " 0  4  <span class='underline'>14</span>  5 ", "parseStack": "T+i", "inStrStack": "*i#", "action": "REDUCE F->i"}
, {"statusStack": " 0  4  <span class='underline'>14</span>  3 ", "parseStack": "T+F", "inStrStack": "*i#", "action": "SHIFT"}
, {"statusStack": " 0  4  <span class='underline'>14</span>  3  <span class='underline'>11</span> ", "parseStack": "T+F*", "inStrStack": "i#", "action": "SHIFT"}
, {"statusStack": " 0  4  <span class='underline'>14</span>  3  <span class='underline'>11</span>  5 ", "parseStack": "T+F*i", "inStrStack": "#", "action": "REDUCE F->i"}
, {"statusStack": " 0  4  <span class='underline'>14</span>  3  <span class='underline'>11</span>  <span class='underline'>25</span> ", "parseStack": "T+F*F", "inStrStack": "#", "action": "REDUCE S->@"}
, {"statusStack": " 0  4  <span class='underline'>14</span>  3  <span class='underline'>11</span>  <span class='underline'>25</span>  <span class='underline'>34</span> ", "parseStack": "T+F*FS", "inStrStack": "#", "action": "REDUCE S->*FS"}
, {"statusStack": " 0  4  <span class='underline'>14</span>  3  <span class='underline'>13</span> ", "parseStack": "T+FS", "inStrStack": "#", "action": "REDUCE T->FS"}
, {"statusStack": " 0  4  <span class='underline'>14</span>  <span class='underline'>27</span> ", "parseStack": "T+T", "inStrStack": "#", "action": "REDUCE G->@"}
, {"statusStack": " 0  4  <span class='underline'>14</span>  <span class='underline'>27</span>  <span class='underline'>36</span> ", "parseStack": "T+TG", "inStrStack": "#", "action": "REDUCE G->+TG"}
, {"statusStack": " 0  4  <span class='underline'>16</span> ", "parseStack": "TG", "inStrStack": "#", "action": "REDUCE E->TG"}
, {"statusStack": " 0  2 ", "parseStack": "E", "inStrStack": "#", "action": "ACCEPT"}]
}
